plugins/arrow-flight-rpc/src/main/java/org/opensearch/arrow/flight/impl/BaseFlightProducer.java

```diff
@@ -0,0 +1,186 @@
+import org.apache.arrow.flight.FlightEndpoint;
+import org.apache.arrow.flight.FlightInfo;
+import org.apache.arrow.flight.Location;
+import org.apache.arrow.flight.NoOpFlightProducer;
+import org.apache.arrow.flight.Ticket;
+import org.apache.arrow.memory.BufferAllocator;
+import org.apache.arrow.vector.VectorSchemaRoot;
+import org.apache.logging.log4j.LogManager;
+import org.apache.logging.log4j.Logger;
+import org.opensearch.arrow.flight.bootstrap.FlightClientManager;
+import org.opensearch.arrow.spi.StreamProducer;
+import org.opensearch.arrow.spi.StreamTicket;
+
+import java.io.IOException;
+import java.util.Collections;
+import java.util.Optional;
+
+/**
+ * BaseFlightProducer extends NoOpFlightProducer to provide stream management functionality
+ * for Arrow Flight in OpenSearch. This class handles the retrieval and streaming of data
+ * based on provided tickets, managing backpressure, and coordinating between the stream
+ * provider and the server stream listener.
+ */
+public class BaseFlightProducer extends NoOpFlightProducer {
+    private final FlightClientManager flightClientManager;
+    private final FlightStreamManager streamManager;
+    private final BufferAllocator allocator;
+    private static final Logger logger = LogManager.getLogger(BaseFlightProducer.class);
+
+    /**
+     * Constructs a new BaseFlightProducer.
+     *
+     * @param flightClientManager The FlightClientManager to handle client connections.
+     * @param streamManager The StreamManager to handle stream operations, including
+     *                      retrieving and removing streams based on tickets.
+     * @param allocator The BufferAllocator for memory management in Arrow operations.
+     */
+    public BaseFlightProducer(FlightClientManager flightClientManager, FlightStreamManager streamManager, BufferAllocator allocator) {
+        this.flightClientManager = flightClientManager;
+        this.streamManager = streamManager;
+        this.allocator = allocator;
+    }
+
+    /**
+     * Handles the retrieval and streaming of data based on the provided ticket.
+     * This method orchestrates the entire process of setting up the stream,
+     * managing backpressure, and handling data flow to the client.
+     *
+     * @param context The call context (unused in this implementation)
+     * @param ticket The ticket containing stream information
+     * @param listener The server stream listener to handle the data flow
+     */
+    @Override
+    public void getStream(CallContext context, Ticket ticket, ServerStreamListener listener) {
+        StreamTicket streamTicket = streamManager.getStreamTicketFactory().fromBytes(ticket.getBytes());
+        Optional<FlightStreamManager.StreamProducerHolder> streamProducerHolder = Optional.empty();
+        try {
+            if (streamTicket.getNodeId().equals(flightClientManager.getLocalNodeId())) {
+                streamProducerHolder = streamManager.removeStreamProducer(streamTicket);
```

Comment:
```
streamProducerHolder = streamManager.removeStreamProducer(streamTicket).orElseThrow(() -> CallStatus.NOT_FOUND.withDescription("Stream not found").toRuntimeException());
```

The listener could be called only when `FlightRuntimeException` happens, the logic is quite entangled with exceptions / listeners and return calls, difficult to follow (at least to me):

try {
 ...
} catch(FlightRuntimeException ex) {
    listener.error(ex);
    throw ex;
} ....
```

Reply:
let me work on this logic here

Reply:
@reta I have improved the code flow and error handling, please let me know what do you think. 
also i have added the changes on top of https://github.com/opensearch-project/OpenSearch/pull/17738, so you will see extra diff. Thank you